### PR TITLE
[DOC] [WIP] Developers documentation page

### PR DESCRIPTION
**Pull request by mvargas33**
@bellet @perimosocordiae @terrytangyuan @wdevazelhes 

Motivated by #259 , I made this docs for new developers like a guide in _How to_ contribute to the package. I followed the scikit-learn guideline [here](https://scikit-learn.org/stable/developers/contributing.html), but talking with @bellet it's better to keep things simple in terms of governance, for instance.

I also considered comments at #205 and in #13 .

I also propose that for API or major changes, a MLEP (Metric Learning Enhancement Proposal) document is needed, being [Github Discussions](https://github.com/scikit-learn-contrib/metric-learn/discussions) the palce to put it and review it, because sometimes, huge API changes are linked to more than one PR. Take the OASIS discussion at #336 as a very simple and informal MLEP. (In general I took this idea from sklearn).

The main sections are:

- **Contributing**: Introduction, values, general guideline, how to contribute code (PR process), how to test, how to compile the docs.
- **Metric learn governance**: Roles, decision-making process, in general. How to proceed with API changes (MLEP)
- **Implement a new algorithm**: Criteria of selection, and how to proceed.
- **API Structure**: A quick review of the API for devs to know what classes to inherit from, and wich methods to implement, and where.
- **MLEP Template**: The template to be used in Github discussion for major changes.

What is left:

- How to make a release
- How to publish the docs (the gh-page branch thing)
- How to update at Pypi and Conda.

And because this has not been discussed in the past, take this draft as what it is: a draft. Moreover the governance part, and the MLEP part. Maybe something much simpler is enough, like the Github discussion #336 that I did, but with a general template.

Best! 😁

Ps: I'm testing CSS usage in some parts, ignore it

### CODE REVIEW

**Comment by mvargas33** [comment]
> Great to see the formal process documented. Let me know when this is ready for review.

Hello! The branch is ready for a review. I added a last section called "Publishing Releases" that has all the information explained at #250 and #259.

I emphasize that **this documentation should be treated as a draft**, as many parts of the contributing process is not yet defined. This PR is an opportunity to clarify it and reach consensus in _How to collaborate_.

It would be great if @perimosocordiae, @wdevazelhes and @bellet could also review this to have a broad opinion about it.

Best! 🧙‍♂️
Max

Ps: I suggest to merge #338 first to remove all warnings when building this PR.

File: doc/contrib/contributing.rst [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+Contributing code
+=================
+
+.. note::
+
+  To avoid duplicating work, it is highly advised that you search
+  through the issue tracker and the PR list. If in doubt about duplicated
+  work, or if you want to work on a non-trivial feature, it’s recommended
+  to first open an issue in the issue tracker to get some feedbacks from
```
Comment (perimosocordiae) [inline doc/contrib/contributing.rst:145]: minor nitpick: feedbacks -> feedback

File: doc/contrib/contributing.rst [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+  core developers.
+
+  One easy way to find an issue to work on is by applying the “help wanted”
+  label in your search. This lists all the issues that have been unclaimed
+  so far. In order to claim an issue for yourself, please comment exactly
+  `take` on it for the CI to automatically assign the issue to you.
```
Comment (perimosocordiae) [inline doc/contrib/contributing.rst:151]: Does this actually work? I've never seen this before.

File: doc/contrib/contributing.rst [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+6. If you need to raise a `Warning`, do a test that verifies that
+   the warning is being shown. Same for `Errors`. Some examples might
+   be warnings about a default configuration, a wrong input, etc.
+
+.. _building-the-docs:
+
+Building the docs
+^^^^^^^^^^^^^^^^^
+
+To build the docs is always recommended to start with a fresh virtual
```
Comment (perimosocordiae) [inline doc/contrib/contributing.rst:352]: is -> it is

File: doc/contrib/release.rst [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+This section has the information needed to do different tasks related to
+releases. This task is usually performed by core developers but it
+might be requested to other developers in the future.
+
+Before the release
+==================
+
+1. Make sure that all versions numbers are updated in ``metric_learn/_version.py```
+   and in ``doc/conf.py```.
+2. Make sure that the final year date in doc/conf.py after copytight is the right
```
Comment (perimosocordiae) [inline doc/contrib/release.rst:16]: typo: copyright

File: doc/preprocessor.rst [hunk 5]
```diff
@@ -4,9 +4,11 @@
 Preprocessor
 ============
 
-Estimators in metric-learn all have a ``preprocessor`` option at instantiation.
-Filling this argument allows them to take more compact input representation
-when fitting, predicting etc...
+.. rst-class:: deprecated
```
Comment (perimosocordiae) [inline doc/preprocessor.rst:7]: Are we still deprecating this? It feels a bit out of place for this PR.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]
